rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -0,0 +1,156 @@
+            )
+            relations = migrate_role(self.role, False, str(self.root_workspace.uuid), self.org_id, True, True, False)
+            self.current_role_relations = relations
+        except Exception as e:
+            raise DualWriteException(e)
+
+    def regenerate_relations_and_mappings_for_role(self):
+        """Delete and generated relations with mapping for a role."""
+        if not self.replication_enabled():
+            return []
+        self.delete_mappings()
+        return self.generate_relations_and_mappings_for_role()
+
+    def generate_relations_and_mappings_for_role(self):
+        """Generate relations and mappings for a role with new UUIDs for v2 role and role bindings."""
+        if not self.replication_enabled():
+            return []
+        try:
+            logger.info("[Dual Write] Generate new relations from role(%s): '%s'", self.role.uuid, self.role.name)
+            relations = migrate_role(self.role, False, str(self.root_workspace.uuid), self.org_id)
+            self.role_relations = relations
+            return relations
+        except Exception as e:
+            raise DualWriteException(e)
+
+    def get_current_role_relations(self):
+        """Get current roles relations."""
+        return self.current_role_relations
+
+    def delete_mappings(self):
+        """Delete mappings for a role."""
+        if not self.replication_enabled():
+            return
+        try:
+            logger.info("[Dual Write] Delete mappings for role(%s): '%s'", self.role.uuid, self.role.name)
+            v2_roles_ids = BindingMapping.objects.filter(v1_role=self.role.id).values("v2_role_id")
+            # this deletes also records in BindingMapping table and role binding
+            V2Role.objects.filter(id__in=v2_roles_ids).delete()
+        except Exception as e:
+            raise DualWriteException(e)
+
+    def set_role(self, role):
+        """Set a role."""
+        self.role = role
+
+    def build_replication_event(self):
+        """Build replication event."""
+        if not self.replication_enabled():
+            return {}
+        logger.info("[Dual Write] Build Replication event for role(%s): '%s'", self.role.uuid, self.role.name)
+        relations_to_add = []
+        for relation in self.role_relations:
+            relations_to_add.append(relationship_to_json(relation))
+
+        relations_to_remove = []
+        for relation in self.current_role_relations:
+            relations_to_remove.append(relationship_to_json(relation))
+
+        replication_event = {"relations_to_add": relations_to_add, "relations_to_remove": relations_to_remove}
+        return replication_event
+
+    def generate_replication_event_to_outbox(self, role):
+        """Generate replication event to outbox table."""
+        if not self.replication_enabled():
+            return
+        self.set_role(role)
```

Comment:
Why does this method accept a `role` if the `role` is already passed when it is initialized? You could drop that and the `set_role` method I think, which makes this class less error prone. For example you could technically pass a `role` with a different tenant, and then `self.tenant_id` etc would be out of sync with `self.role`.

Reply:
I need new state of `role` but I can remove `set_role` method